shared/src/ReferenceCounter.cs

```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Threading;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// A class that is used to handle an object that needs to be counted.
+    /// </summary>
+    /// <remarks>
+    /// Use of this class should be a static member variable inside of another class.
+    /// </remarks>
+    /// <example>
+    /// <code> public class ContainerUsage
+    /// {
+    ///     private ReferenceCounter&lt;object&gt; _refObject = new ReferenceCounter&lt;object&gt;();
+    ///     
+    ///     public ContainerUsage()
+    ///     {
+    ///         _refObject.Create(() => new object());
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    /// <typeparam name="T">Any class based object type.</typeparam>
+    public class ReferenceCounter<T> where T : class
+    {
+        private T _obejctToCount;
+
+        private object _lockObject = new object();
+
+        private Action<T> _executeOnDispose;
+
+        private volatile int _referenceCount;
+
+        /// <summary>
+        /// The reference counted object.
+        /// </summary>
+        public T Value => _obejctToCount;
+
+        /// <summary>
+        /// The current count.
+        /// </summary>
+        public int Count => _referenceCount;
+
+        /// <summary>
+        /// Create the object or increment the counter.
+        /// </summary>
+        /// <remarks>
+        /// If the object is already set this will not execute the code in <paramref name="objectToCreate"/>. Instead it will return the object that was previously created.
+        /// 
+        /// 
+        /// 
+        /// This will return a mutable object. It is possible to alter the object state which could cause unexpected behavior.
+        /// </remarks>
+        /// <param name="objectToCreate">A function that generates the object to be counted. This must not return a null object.</param>
+        /// <exception cref="InvalidOperationException">This will be thrown if the reference object is somehow out of sync.</exception>
+        /// <exception cref="ArgumentException">This will be thrown if the creation method supplied to this method returns a null object.</exception>
+        /// <returns>The object created</returns>
+        public T Create(Func<T> objectToCreate)
+        {
+            lock (_lockObject)
+            {
+                CreateWithoutLocking(objectToCreate);
+                Interlocked.Increment(ref _referenceCount);
+                return _obejctToCount;
+            }
+        }
+
+        private void CreateWithoutLocking(Func<T> objectToCreate)
+        {
+            if (_obejctToCount == null && _referenceCount <= 0)
```

Comment:
under what circumstances does `_referenceCount` go below zero?

Reply:
None, now that think of it.